### DISCUSSION TITLE
Any recommended strategies for cache invalidation

### DISCUSSION THREAD

**Discussion by michaelbukachi** [sqlalchemy/dogpile.cache#218] [Usage Questions]
This question might not be specific to this library but just a general question. Are there any recommended approaches for cache invalidation when relationships (related data) are involved? 
E.g
```python
class User(Base):
     id = Column(...)
    name = Column(...)

    roles = relationship('Role', ....)

class Role(Base)
     id = Column(....)
    name = Column(....)
    user_id = Column(....)
```
How would we go about invalidating a user when we update one of their roles? I'm able to invalidate all users when a role is updated, however, I don't think that is ideal. I'm using the [official example](https://github.com/sqlalchemy/sqlalchemy/tree/main/examples/dogpile_caching) as a reference.

**Note**: Assume no lazyloading is being used i.e only eagerloading

**Comment by zzzeek** [comment]
cc @jvanasco , IIRC he has a lot of opinions on this

**Reply by michaelbukachi** [reply]
@jvanasco please share your opinions 😁. Any videos, articles etc?

**Reply by jvanasco** [reply]
There's no universal solution to caching and invalidation.  Getting it right on one project is wrong for another.

In very general terms, you want to have a good understanding of what your cached data is, how it is structured, and the access patterns.  When it comes to User/Role situations like your example, I generally view that association as (i) Very High Read, (ii) Low Write, and (iii) Strongly Partitioned from the User and Role records in the sense that I don't need any of the record's other information - just the existence or lack-of the association.  In the context of a cloud based web application, my goal is to load the user's role information on every page request or action.  With a scenario like that, I would likely have a dogpile "region" named/prefixed `user2role` with the key as the UserID and the value as a list of the role ids.  That would allow me to efficiently store and query the association data, and avoid the overhead of storing/transferring user profile or role information. 

When it comes to invalidation, I like to create triggers and policies.  If I edit a User's SQLAlchemy model (in the database), I will either catch that with a SQLAlchemy Event to automatically trigger a policy, or I will explicitly invoke a a policy.  The policy is just a small routine that says "If the edited object is a user, invalidate these cache items; if they edited object is a user and a role, invalidate these cache items as well".

**Reply by michaelbukachi** [reply]
@jvanasco You are totally right. There is truly no universal solution 😱. I think I figured out an implementation that works though. I was struggling with coming with the right policies i.e when this obj changes, invalidate this obj and that obj as well. Thanks for the insights though, I had totally overlooked "partitioning".
